### ISSUE TITLE
Hybrid predictions using RF

### ISSUE THREAD

**Issue by vishwas-gpb** [vishwas-gpb/Hybrid-prediction-using-ensemble-machine-learning#1]
`library(readxl)
library(dplyr)
library(randomForest)
library(ggplot2)
library(gridExtra)
library(grid)

# Load data from Excel
file_path <- read.csv("RF featrures.csv")
df <- file_path

# List of traits extracted from column naming pattern for parent trait means (e.g., DFF, Yield, etc.)
# Extract traits by removing season and suffix; assuming unified traits
all_cols <- colnames(df)

# Extract traits from parent mean columns e.g. 'S1_DFF_Mean', pattern to pick trait names
traits_s1 <- gsub("S1_|_Mean", "", grep("^S1_.*_Mean$", all_cols, value = TRUE))
traits_s2 <- gsub("S2_|_Mean", "", grep("^S2_.*_Mean$", all_cols, value = TRUE))
traits <- unique(c(traits_s1, traits_s2))

# Features often used:
# ClusterDiff, GeneticDist
# All P1_G_* and P2_G_* columns for GCA effects
# Parent seasonal mean trait columns: S1_trait_Mean, S2_trait_Mean

# Loop for each trait to train and predict with random forest
set.seed(2024)
for (trait in traits) {
  # Build vector of feature column names for this trait
  parent_mean_cols <- c(paste0("S1_", trait, "_Mean"), paste0("S2_", trait, "_Mean"))
  
  # Construct predictor columns manually for GCA columns that start with P1_G_ and P2_G_ (all)
  P1_G_cols <- grep("^P1_G_", all_cols, value = TRUE)
  P2_G_cols <- grep("^P2_G_", all_cols, value = TRUE)
  
  # Core features
  feature_cols <- c("ClusterDiff", "GeneticDist", P1_G_cols, P2_G_cols, parent_mean_cols)
  
  # Filter only existing columns (in case some not present)
  feature_cols <- feature_cols[feature_cols %in% all_cols]
  
  # Dependent variable (response)
  resp_col <- paste0(trait, "_Hybrid")
  if (!(resp_col %in% all_cols)) {
    warning(paste("Response column", resp_col, "not found. Skipping trait:", trait))
    next
  }
  
  # Prepare predictors and response
  X <- df[, feature_cols]
  y <- df[[resp_col]]
  
  # Train random forest model
  rf_model <- randomForest(X, y)
  
  # Predict on training data (can later do test set prediction)
  df[[paste0("Pred_", trait)]] <- predict(rf_model, X)
  
  # Calculate prediction error/difference
  df[[paste0("Diff_", trait)]] <- df[[paste0("Pred_", trait)]] - y
}

# Export predictions and differences
write.csv(df, "hybrid_predictions_extended.csv", row.names = FALSE)

# Plot observed vs predicted and errors for each trait
obs_pred_plots <- list()
error_plots <- list()

for(trait in traits) {
  resp_col <- paste0(trait, "_Hybrid")
  pred_col <- paste0("Pred_", trait)
  diff_col <- paste0("Diff_", trait)
  
  if (!(resp_col %in% colnames(df)) || !(pred_col %in% colnames(df))) next
  
  p1 <- ggplot(df, aes(x = .data[[resp_col]], y = .data[[pred_col]])) +
    geom_point(color = "steelblue", size = 3, alpha = 0.7) +
    geom_abline(slope = 1, intercept = 0, linetype = "dashed", color = "darkred") +
    labs(title = trait, x = "Observed", y = "Predicted") +
    theme_minimal() +
    theme(plot.title = element_text(size = 14, face = "bold", hjust = 0.5))
  obs_pred_plots[[trait]] <- p1
  
  p2 <- ggplot(df, aes(x = factor(1:nrow(df)), y = .data[[diff_col]])) +
    geom_bar(stat = "identity", fill = "tomato", color = "black", width = 0.7) +
    labs(title = trait, x = "Hybrid Index", y = "Prediction Error") +
    theme_minimal() +
    theme(plot.title = element_text(size = 14, face = "bold", hjust = 0.5))
  error_plots[[trait]] <- p2
}

# Display plots
grid.arrange(arrangeGrob(grobs = obs_pred_plots, ncol = 2), top = textGrob("Observed vs Predicted Hybrid Performance", gp = gpar(fontsize = 16, fontface = "bold")))
grid.arrange(arrangeGrob(grobs = error_plots, ncol = 2), top = textGrob("Prediction Errors by Hybrid", gp = gpar(fontsize = 16, fontface = "bold")))
`
